### PR TITLE
Soil nutrient concentration bugfix

### PR DESCRIPTION
**Pull request by qzhu-lbl**
- soil nh4 solution concentration unit conversion bugfix
 - soil p concentration initialization bugfix
 - nutrient and water scalar bugfix

BFB for CTC compsets, non-BFB for ECA compsets

### CODE REVIEW

**Comment by jinyuntang** [comment]
@qzhu-lbl, this looks good to me. BTW, given the fix, should the parameter file be changed as well?

**Comment by rgknox** [comment]
@qzhu-lbl , there should be significant conflicts generated from PR #4325, but most of the key changes are included there as well:

Unit conversion for available ammonium:
https://github.com/E3SM-Project/E3SM/pull/4325/files#diff-f9276b0713e2af1bcd0495bdbd37de680e27b15dcf180b1d8da1ec0243ba17dfR2976

decompmicc should be calculated and ready for use in all relevant scenarios:
https://github.com/E3SM-Project/E3SM/pull/4325/files#diff-f9276b0713e2af1bcd0495bdbd37de680e27b15dcf180b1d8da1ec0243ba17dfR1152

The calculations of dsolutionp_dt should also be in the right place, calculated AFTER plant demand for P is calculated:
https://github.com/E3SM-Project/E3SM/pull/4325/files#diff-f9276b0713e2af1bcd0495bdbd37de680e27b15dcf180b1d8da1ec0243ba17dfR3318

Still need to to swap out the cn_scalar and cp_scalar for the "runmean" version though

**Comment by bishtgautam** [comment]
@qzhu-lbl As @rgknox predicted, there are a lot of conflicts after #4325 was merged. Could you please rebase your branch?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:109]: `decompmicc`
  - In `master` is 1D for `j`-level
  - In this branch is 2D for `c,j`
Question for @qzhu-lbl @rgknox: Should this be 1D or 2D?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1195]: @qzhu-lbl @rgknox It appears this has already been fixed in master. Correct?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1204]: @qzhu-lbl @rgknox It appears this has already been fixed in master. Correct?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1244]: @qzhu-lbl @rgknox These fixes are needed in master. Correct?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1526]: @qzhu-lbl @rgknox This fix is needed in master

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1551]: @qzhu-lbl @rgknox It appears this change is already in master. Correct?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1884]: @qzhu-lbl @rgknox These fixes are needed in the master. Correct?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:2348]: @qzhu-lbl @rgknox It appears this change is needed in master. Correct?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:2521]: @qzhu-lbl @rgknox It appears this fix is needed in master. Correct?

**Review by bishtgautam — COMMENTED** [review]
@qzhu-lbl @rgknox While trying to rebase this branch there are major conflicts in `AllocationMod.F90`. I have made a few comments on whether the changes in `AllocationMod.F90` made in this PR are needed or not, but I need your help in reviewing my comments.

cc: @susburrows @jenniferholm @rljacob

**Comment by rgknox** [comment]
I think we should just manually re-add the elements in this PR onto master, my PR with the nutrient refactor incorporates most of this PR.  @qzhu-lbl , do you want me to try this and check it with you?

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:2521]: yes

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:2348]: yes

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1884]: yes

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1551]: yes, should be already in

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1526]: yes

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1244]: yes

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1204]: yes, already in

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1195]: yes this is in, but @qzhu-lbl  should have the final word on what this new scalar value should be... I think we are both using 1e-3 ....

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:109]: asfaict (as far as I can tell, "as far as I can tell" is not yet an acronym in the public lexicon, so I will define it here, asfaict = as far as I can tell) there is no reason to make this 2D, because we do not need to remember the values from other columns as we make calculations on the column of interest.  We also don't output this in history, so allocating just on soil depth is cheaper and more efficient.

**Comment by bishtgautam** [comment]
@rgknox I will manually add the changes, but I want to confirm that I have correctly identified all the changes.

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (bishtgautam) [inline components/elm/src/biogeochem/AllocationMod.F90:1143]: @rgknox It appears that FATES does not have `<cn/cp>_scalar_runmean`, so I have set them to `<cn/cp>_scalar`.

Comment (rgknox) [reply components/elm/src/biogeochem/AllocationMod.F90:1143]: It also looks like we don't need the _scalar_ptr's anymore.

**Review by rgknox — APPROVED** [review]

**Comment by bishtgautam** [comment]
@qzhu-lbl This PR is not BFB for the crop model. See the DIFFs for `IELMCNCROP` tests in https://my.cdash.org/viewTest.php?onlyfailed&buildid=2090834. Can you fix this PR for the crop model? If you can't fix the PR by late afternoon today, I will have to revert this PR from next.

cc: @susburrows @jenniferholm 
fyi: @bbye

**Comment by bishtgautam** [comment]
@qzhu-lbl Disregard my previous comment as it turns out the DIFFs in the crop model are due to new fields being added in the history file. Sorry for the false alarm.

File: components/elm/src/biogeochem/AllocationMod.F90
Comment (qzhu-lbl) [reply components/elm/src/biogeochem/AllocationMod.F90]: @bishtgautam @rgknox yes, 1e-3 is the correct scalar to use.

**Comment by jenniferholm** [comment]
@rgknox , @qzhu-lbl : I'm a little late on looking at this PR, but glad to see it's merged into master.  Can you guys confirm or not, do these changes impact FATES nutrient modeling in anyway? Like are any of the updated variables here connected to FATES, or is everything local to ELM?  For example, it looks like you pointed out that FATES does not have the cp/cn_scalar_runmean_prt variable...

**Pull request merged**
